src/core/hrs.ml

```diff
@@ -110,8 +115,11 @@ let to_HRS : Format.formatter -> Sign.t -> unit = fun oc sign ->
   Format.fprintf oc "\n(COMMENT rewriting rules)\n";
   let print_rules s =
     match !(s.sym_def) with
-    | Some(d) -> Format.fprintf oc "(RULES %a\n" print_sym s;
-                 Format.fprintf oc "    -> %a)\n" (print_term false) d
-    | None    -> List.iter (print_rule oc s) !(s.sym_rules)
+    | Some(d) -> print_rule oc (Symb(s,Nothing)) d
+    | None    ->
+      let rules = !(s.sym_rules) in
+      let lhs r = Basics.add_args (Symb(s,Qualified)) r.lhs in
+      let rhs r = Basics.term_of_rhs r in
+      List.iter2 (print_rule oc) (List.map lhs rules) (List.map rhs rules)
```

Comment:
I don't really like what you have done here, since this leads to several traversal of the list of rules in the `None` case (`List.iter` and `List.map` twice). I think that a better solution would be to keep the same interface as before for `print_rule`, but introduce a new intermediate function (exactly your `print_rule` in fact, but it should be renamed to `print_rule_aux` for example). This way the intermediate function can be called in the `Some` case directly, and in the `None` case indirectly through `print_rule`.